osutil/io.go

```diff
@@ -231,44 +232,57 @@ func AtomicWriteChown(filename string, reader io.Reader, perm os.FileMode, flags
 func AtomicRename(oldName, newName string) error {
 	var oldDir, newDir *os.File
 
+	// snapdUnsafeIO controls the ability to ignore expensive disk
+	// synchronization. It is only used inside tests.
 	if !snapdUnsafeIO {
 		oldDirPath := filepath.Dir(oldName)
 		newDirPath := filepath.Dir(newName)
 
-		o, err := os.Open(oldDirPath)
+		oldDir, err := os.Open(oldDirPath)
 		if err != nil {
 			return err
 		}
-		oldDir = o
-		defer o.Close()
+		defer oldDir.Close()
 
-		if oldDir != newDir {
-			n, err := os.Open(newDirPath)
-			if err != nil {
-				return err
+		newDir, err := os.Open(newDirPath)
+		if err != nil {
+			return err
+		}
+		defer newDir.Close()
+
+		oldInfo, err := oldDir.Stat()
+		if err != nil {
+			return err
+		}
+		newInfo, err := newDir.Stat()
+		if err != nil {
+			return err
+		}
+		if oldStat, ok := oldInfo.Sys().(*syscall.Stat_t); ok {
+			if newStat, ok := newInfo.Sys().(*syscall.Stat_t); ok {
+				// Old and new directories refer to the same location. We can only sync once.
+				if oldStat.Dev == newStat.Dev && oldStat.Ino == newStat.Ino {
```

Comment:
This is a good test to have, having both entries in the same directory can be the typical case.